File: app/jenkins/src/main/scala/com/kstreee/ci/app/JenkinsApp.scala
package com.kstreee.ci.app

import com.kstreee.ci.analysis.AnalysisConfig
import hudson.model.{BuildListener, FreeStyleBuild}
import hudson.util.FormValidation
import hudson.{Extension, Launcher}
import net.sf.json.JSONObject
import org.kohsuke.stapler.{DataBoundConstructor, QueryParameter, StaplerRequest}

import scala.beans.BeanProperty
import scala.concurrent.Future

object JenkinsApp extends App {
  import scala.concurrent.ExecutionContext.Implicits.global

  override type T = String

  override def loadConfig(args: String): Future[Option[AnalysisConfig]] = {
    Future(None)
  }

  @Extension
  class DescriptorScalaImpl extends BuildDescriptorScalaAdapter {
    @BeanProperty
    var useFrench: Boolean = _

    def doCheckName(@QueryParameter value: String): FormValidation = {
      if (value.length == 0) return FormValidation.error("Please set a name")
      if (value.length < 4) return FormValidation.warning("Isn't the name too short?")
      FormValidation.ok()
    }

    override def getDisplayName: String = "Say hello world"

    override def configure(req: StaplerRequest, formData: JSONObject): Boolean = {
      useFrench = formData.getBoolean("useFrench")
      super.configure(req, formData)
    }
  }
}

@DataBoundConstructor
class JenkinsApp(@BeanProperty val name: String) extends BuilderScalaAdapter {
  override def scala_perform(build: FreeStyleBuild, launcher: Launcher, listener: BuildListener): Boolean = {
    if (getDescriptor.getUseFrench) listener.getLogger.println("Bonjour, " + name + "!") else listener.getLogger.println("Hello, " + name + "!")
    true
  }

  override def getDescriptor: JenkinsApp.DescriptorScalaImpl = {
    super.getDescriptor.asInstanceOf[JenkinsApp.DescriptorScalaImpl]
  }
}

File: app/jenkins/src/main/java/com/kstreee/ci/app/JenkinsApp.java
package com.kstreee.ci.app;

import hudson.Extension;
import hudson.FilePath;
import hudson.Launcher;
import hudson.model.AbstractProject;
import hudson.model.Run;
import hudson.model.TaskListener;
import hudson.tasks.BuildStepDescriptor;
import hudson.tasks.Builder;
import hudson.util.FormValidation;
import java.util.concurrent.Executors;
import javax.annotation.Nonnull;
import jenkins.tasks.SimpleBuildStep;
import org.kohsuke.stapler.DataBoundConstructor;
import org.kohsuke.stapler.QueryParameter;


public class JenkinsApp extends Builder implements SimpleBuildStep {
  private final String name;

  @DataBoundConstructor
  public JenkinsApp(String name) {
    this.name = name;
  }

  public String getName() {
    return name;
  }

  @Override
  public void perform(@Nonnull Run<?,?> run,
                      @Nonnull FilePath workspace,
                      @Nonnull Launcher launcher,
                      @Nonnull TaskListener listener) {
    try {
      listener.getLogger().println("Hello, " + name + "!");
      AppJavaCompatible.analysisJavaCompatible(null, Executors.newFixedThreadPool(1)).wait();
    } catch (Exception e) {
      listener.getLogger().flush();
    }
  }

  @Extension
  public static final class DescriptorImpl extends BuildStepDescriptor<Builder> {
    public FormValidation doCheckName(@QueryParameter String value,
                                      @QueryParameter boolean useFrench) {
      return FormValidation.ok();
    }

    @Override
    public boolean isApplicable(Class<? extends AbstractProject> clazz) {
      return true;
    }

    @Override
    public String getDisplayName() {
      return "hello world";
    }
  }
}